sdks/csharp/adtIotHub_allocate.cs

```diff
@@ -65,75 +67,83 @@ public static class DpsAdtAllocationFunc
 
-            return (fail)
+            return fail
                 ? new BadRequestObjectResult(message)
-                : (ActionResult)new OkObjectResult(obj);
+                : (ActionResult)new OkObjectResult(response);
         }
 
-        public static async Task<string> FindOrCreateTwin(string dtmi, string regId, ILogger log)
+        public static async Task<string> FindOrCreateTwinAsync(string dtmi, string regId, ILogger log)
         {
             // Create Digital Twins client
             var cred = new ManagedIdentityCredential(adtAppId);
-            var client = new DigitalTwinsClient(new Uri(adtInstanceUrl), cred, new DigitalTwinsClientOptions { Transport = new HttpClientTransport(httpClient) });
+            var client = new DigitalTwinsClient(
+                new Uri(adtInstanceUrl),
+                cred,
+                new DigitalTwinsClientOptions
+                {
+                    Transport = new HttpClientTransport(singletonHttpClientInstance)
+                });
 
             // Find existing twin with registration ID
-            string dtId;
             string query = $"SELECT * FROM DigitalTwins T WHERE $dtId = '{regId}' AND IS_OF_MODEL('{dtmi}')";
-            AsyncPageable<string> twins = client.QueryAsync(query);
+            AsyncPageable<BasicDigitalTwin> twins = client.QueryAsync(query);
+            string dtId;
 
-            await foreach (string twinJson in twins)
+            await foreach (BasicDigitalTwin digitalTwin in twins)
             {
-                // Get DT ID from the Twin
-                JObject twin = (JObject)JsonConvert.DeserializeObject(twinJson);
-                dtId = (string)twin["$dtId"];
+                dtId = digitalTwin.Id;
                 log.LogInformation($"Twin '{dtId}' with Registration ID '{regId}' found in DT");
-                return dtId;
+                break;
             }
 
-            // Not found, so create new twin
-            log.LogInformation($"Twin ID not found, setting DT ID to regID");
-            dtId = regId; // use the Registration ID as the DT ID
-
-            // Define the model type for the twin to be created
-            Dictionary<string, object> meta = new Dictionary<string, object>()
-            {
-                { "$model", dtmi }
-            };
-            // Initialize the twin properties
-            Dictionary<string, object> twinProps = new Dictionary<string, object>()
+            if (String.IsNullOrWhiteSpace(dtId))
             {
-                { "$metadata", meta }
-            };
-            twinProps.Add("Temperature", 0.0);
+                // Not found, so create new twin
+                log.LogInformation($"Twin ID not found - setting DT ID to regID");
+                dtId = regId; // use the Registration ID as the DT ID
+
+                // Initialize the twin properties
+                var digitalTwin = new BasicDigitalTwin
```

Comment:
We specify `BasicDigitalTwin` in our type below, but weren't using it here. So I updated the initialization code to use that type explicitly. I think it reads better too. Thoughts?